Initial implementation of a cactus stack.

ltratt opened the pull request:
An immutable cactus stuck (also called a spaghetti stack or parent pointer tree). A cactus stack is a (possibly empty) node with a (possibly null) pointer to a parent node. Pushing and popping return new nodes rather than mutating existing nodes.

```rust
use cactus::Cactus;
let c = Cactus::new();
assert!(c.is_empty());
let c2 = c.push(1);
assert_eq!(c2.len(), 1);
assert_eq!(*c2.val().unwrap(), 1);
let c3 = c2.pop().unwrap();
assert!(c3.is_empty());
```

From a given node one can create multiple sub-stacks:

```rust
use cactus::Cactus;
let c = Cactus::new().push(1);
let c2 = c.push(2);
let c3 = c.push(3);
assert!(c2 != c3);
assert_eq!(c2.iter().cloned().collect::<Vec<_>>(), [2, 1]);
assert_eq!(c3.iter().cloned().collect::<Vec<_>>(), [3, 1]);
```

Notice that the big difference between this and the (unfinished?) Kaktus library is that this one has immutable nodes. That means that we can use `Rc`s on their own without (expensive) `RefCell`s. And yes, it turns out that I do need a cactus stack for another program I'm writing!

jaheba commented:
* I ran `cargo fmt` on the project and the few changes it made seem sensible.

* clippy suggest to remove the lifetimes here `pub fn iter<'a>(&'a self) -> CactusIter<'a, T> {`

ltratt commented:
I've just done (and -- cheating -- force pushed) the Clippy changes.

jaheba commented on src/lib.rs:
```diff
@@ -0,0 +1,272 @@
+    }
+
+    /// How many items are there in this cactus stack?
+    pub fn len(&self) -> usize {
```
I think the whole method body can be replaced with `this.iter().count()`

ltratt replied:
I'm not sure I want to create the iterator just to throw it away, although maybe I'm over-optimising things...

jaheba replied:
I would assume that it's optimised away and in my opinion the iterator version looks much cleaner.

ltratt replied:
I did some (not very high quality) benchmarking. It looks like the iterator version might even be a bit quicker. It's certainly no slower. I've pushed an update accordingly.

vext01 replied:
Neat!

ltratt commented:
[My bad for the force push. Mea culpa.]

jaheba commented on src/lib.rs:
````diff
@@ -0,0 +1,272 @@
+    /// let c3 = c2.push(2);
+    /// assert_eq!(c3.iter().cloned().collect::<Vec<_>>(), [2, 1]);
+    /// ```
+    pub fn push(&self, val: T) -> Cactus<T> {
````
This can be simplified as well

```rust
    pub fn push(&self, val: T) -> Cactus<T> {
        Cactus {
            node: Some(Rc::new(Node {
                val,
                parent: self.node.clone(),
            })),
        }
    }
```

ltratt replied:
Agreed.

jaheba commented on src/lib.rs:
````diff
@@ -0,0 +1,272 @@
+    /// assert_eq!(c2.val(), Some(&1));
+    /// assert_eq!(c2.pop().unwrap(), Cactus::new());
+    /// ```
+    pub fn pop(&self) -> Option<Cactus<T>> {
````
`self.node.as_ref().map(|node| Cactus { node: node.parent.clone() } )`

ltratt replied:
Agreed, that's definitely an improvement.

jaheba commented on src/lib.rs:
````diff
@@ -0,0 +1,272 @@
+    /// assert_eq!(c.iter().cloned().collect::<Vec<_>>(), [3, 2, 1]);
+    /// ```
+    pub fn iter(&self) -> CactusIter<T> {
+        CactusIter{next: self.node.as_ref().map(|n| &*n)}
````
just `self.node.as_ref()` should be sufficient

jaheba commented on src/lib.rs:
```diff
@@ -0,0 +1,272 @@
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.next.take().map(|n| {
+            self.next = n.parent.as_ref().map(|n| &*n);
```
just `self.node.as_ref()` should be sufficient

ltratt replied:
Not sure I understand.

ltratt replied:
Ah, I'm with you.

jaheba commented:
Also, I personally prefer to not use one letter variable names, eg:

`self.node.as_ref().map(|n| &n.val)` -> `self.node.as_ref().map(|node| &node.val)`.

But everything else looks good!

ltratt commented:
OK, I've pushed a commit which addresses most of your comments.

[I politely disagree about single-letter variables in closures. Once upon a time I liked long variable names. Then I realised that the best source code I was reading used shorter names, often single letter when in a tight scope. So since their source code was better than mine, I've switched... Occasionally I lapse into longer variable names but I try not to!]

ltratt commented:
I think I'm done with Jasper's comments. Edd, do you have any? If not, I suggest that this is probably ready for squashing.

vext01 commented on Cargo.toml:
```diff
@@ -0,0 +1,7 @@
+[package]
+name = "cactus"
+version = "0.1.0"
+authors = ["Laurence Tratt <laurie@tratt.net>"]
```
Does rust have the ability to specify a license in the Cargo file?

vext01 replied:
(random question)

ltratt replied:
Yes, I'm working on that stuff separately (probably for the next PR).

vext01 replied:
ok

vext01 commented on src/lib.rs:
```diff
@@ -0,0 +1,250 @@
+
+    /// Is this cactus stack node empty?
+    ///
+    /// # Examples
```
I wonder if there are too many examples. E.g. it's pretty obvious how `is_empty` works.

vext01 commented on src/lib.rs:
```diff
@@ -0,0 +1,250 @@
+        self.iter().count()
+    }
+
+    /// Create a new cactus stack node containing value `val` and parent `self`.
```
And the transitive parents, right?

ltratt replied:
?

vext01 replied:
The new stack will contain the parent of the parent, and so on? I think I got confused by the wording. I would have worded it like:

```
Create a new cactus stack by cloning the existing one and adding `val` as the parent of the current `val`.
```

Or something like that.

ltratt replied:
I agree that the wording here is tricky to get right and I'm all ears for improvements. However, this method doesn't clone the existing stack (it clones the existing stacks internal `Rc` cell, but that's an implementation detail that we want to shield the user from).

vext01 replied:
Doesn't it clone the `Option` containing the `Rc`? That being the case, this ties in with my lifetime question above. There can be sharing between stacks, which invalidates much of my argument, I think.

As for wording, can you see if you can capture the fact that the new stack contains all the old nodes and the one new one? I guess that it what I felt was missing.

ltratt replied:
Yes, it clones the `Option` but that's purely because it's not worth doing `Some(node.unwrap().clone())` (i.e. it's just an optimisation of sorts).

I'll see if I can fix the wording.

vext01 commented on src/lib.rs:
```diff
@@ -0,0 +1,250 @@
+        }
+    }
+
+    /// Return this cactus stack node's parent node or `None` if this cactus stack is empty.
```
I would have expected this method to have returned the item you popped and (because the stack is immutable) a new stack.

I guess it's trivial to get the current item off the top of the stack, but to call this `pop` may be confusing?

Not sure.

ltratt replied:
It could be called `parent` I suppose?

vext01 replied:
Your call. I don't want to come in guns blazing when I'm not really sure myself.

vext01 commented on src/lib.rs:
```diff
@@ -0,0 +1,250 @@
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.next.take().map(|n| {
```
This is not the fault of your code, but  `Option`'s `take()` confuses me. The docs are not helping much either.

```
Takes the value out of the option, leaving a None in its place.
```

But the `Option` is immutable, so how can it put a `None` in? What if the `Option` is already `None`?

ltratt replied:
Notice that `fn take(&mut self)` i.e. `take` can mutate the `self` object.

vext01 replied:
Right, so this is just a convenient way to unwrap an Option?

ltratt replied:
No, it moves the contents out. So if you had `Some(1)` then did `take`, you'd end up with a `None` in place of the `Some(1)`.

vext01 replied:
ok

vext01 commented on src/lib.rs:
```diff
@@ -0,0 +1,250 @@
+    }
+}
+
+impl<T: PartialEq> PartialEq for Cactus<T> {
```
OK, I don't know much about this stuff in Rust, but are you implementing a partial ordering? It seems to me that a stack should be totally ordered?

Maybe that's just the way you have to do it in rust.

ltratt replied:
I'm not sure if/when one also needs `Eq` to be honest. Happy to be enlightened!

vext01 replied:
I'm not sure. The docs only confused me more.

I suggest you go ahead, and if this becomes an issue, fix it later.

jaheba replied:
`PartialEq != PartialOrd`

PartialEq just means that a given value T does not have to be equal to itself (e.g. `NaN != NaN`). In contrast `Eq` signals that the type is also reflexive but doesn't add any other behaviour. If we implement `PartialEq` for our type, we also implement it for `Eq` so this should be fine.

vext01 replied:
I'll take your word for it.

vext01 commented on src/lib.rs:
```diff
@@ -0,0 +1,250 @@
+        if self.len() != other.len() {
+            return false;
+        }
+        self.iter().zip(other.iter()).all(|(x, y)| x == y)
```
nice. (zip)

vext01 commented on src/lib.rs:
```diff
@@ -0,0 +1,250 @@
+    use super::*;
+
+    #[test]
+    fn test_simple() {
```
Would it be better to split this larger test into smaller ones for: e.g. `pop`, `is_empty`, like you did for `iter`?

I guess you could keep the big test too.

ltratt replied:
I'd suggest this is best left as one test.

vext01 replied:
ok

vext01 commented on src/lib.rs:
```diff
@@ -0,0 +1,253 @@
+/// or not.
+#[derive(Clone, Default)]
+pub struct Cactus<T> {
+    node: Option<Rc<Node<T>>>
```
I was trying to figure out why you use an `Rc` here.

It's because the values in the nodes can come from different incompatible lifetimes?

I suspect if you were to clone `val` in push, the `Rc` could go? Not sure, and I'm pretty sure you have tried lots of variations and found this to be the only one you can get rust to accept (?).

vext01 replied:
I mean, in theory, because the stack is immutable, and since there is no sharing when a new stack is made on `push`, it seems to me that all the nodes should have the same lifetime and die together.

vext01 replied:
But then again, you never know with Rust.

ltratt replied:
Because multiple children may point to the same parent so we can't guarantee to free a parent at the same point as a child.

vext01 replied:
Hrm. I'm not going to persist on this, because it's likely I am wrong, but it seems like all of the nodes can die at the same time?

ltratt replied:
I'm not sure how to parse the "can" in that second sentence. Yes they *can* all die at the same time; but they don't *have* to die at the same time. If you think about it: I make a cactus C; push a child C'; I then let C' die; but C could/should be still alive at this point and I can push new children if I want.

vext01 replied:
ok, i managed to convince myself this is probably ok.

vext01 reviewed (commented):
Some comments, but mostly newbie questions.

You guys are much better at Rust than me.

ltratt commented:
I've renamed push and pop: see how you feel about the new versions.

vext01 commented:
At first I wasn't sure, but then I realised that since `pop` may not actually free up a node (someone else may have a reference to it), the notation you proposed actually is quite good.

What do you think though?

ltratt commented:
Well, I *think* the new names are better at the moment. I might wake up tomorrow and change my mind of course ;)

vext01 commented:
It's a tough call. I'll let you choose.

ltratt commented:
I think `child`/`parent` are an improvement and worth sticking with (unless I change my mind tomorrow at which point I can raise a new PR!).

@jaheba Thoughts?

If not, I think we're at squashing time...

jaheba commented:
I personally prefer `push` and `pop`, since these are the operations on a stack. Also `child()` sounds like a getter for me. But it's your call.

ltratt commented:
The problem is that, for an immutable stack, `push` doesn't actually push anything (nor does `pop` pop anything), so I agree with Edd that they're confusing names. I'm not claiming the replacements are perfect though!

ltratt commented:
OK, so unless anyone objects shortly, I'll squash this and propose it to be merged.

ltratt committed:
Initial implementation of a cactus stack.

ltratt commented:
OK, squashed, and (hopefully) ready for merging.

vext01 commented:
> ltratt: 5:54 PM @ebarrett btw, you can feel free to merge the cactus PR i think

Laurie is waiting on this, so let's push on.

vext01 commented:
All ready @ltratt?

jaheba commented:
No objections here

On Fri, 21 Jul 2017 at 18:56, Edd Barrett <notifications@github.com> wrote:

> All ready @ltratt <https://github.com/ltratt>?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/softdevteam/cactus/pull/1#issuecomment-317054940>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAIKZIJJ8d8pLX1nhCrJTwhGGSafv93Qks5sQNgxgaJpZM4Ob1tk>
> .
>

ltratt commented:
I think it's ready.

ltratt commented:
@jaheba You'll need to approve the changes I think before this can be merged.
